File: bot-api/java/src/test/java/dev/robocode/tankroyale/botapi/ConstantsTest.java
package dev.robocode.tankroyale.botapi;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

public class ConstantsTest {

    private static final double EPS = 1e-10;

    @Test
    public void givenDefinedConstants_whenChecked_thenValuesMatchSpec() {
        assertEquals(18, Constants.BOUNDING_CIRCLE_RADIUS);
        assertEquals(1200, Constants.SCAN_RADIUS);
        assertEquals(10, Constants.MAX_TURN_RATE);
        assertEquals(20, Constants.MAX_GUN_TURN_RATE);
        assertEquals(45, Constants.MAX_RADAR_TURN_RATE);
        assertEquals(8, Constants.MAX_SPEED);

        assertEquals(0.1, Constants.MIN_FIREPOWER, EPS);
        assertEquals(3.0, Constants.MAX_FIREPOWER, EPS);

        // Bullet speed by formula
        assertEquals(20 - 3 * Constants.MAX_FIREPOWER, Constants.MIN_BULLET_SPEED, EPS);
        assertEquals(20 - 3 * Constants.MIN_FIREPOWER, Constants.MAX_BULLET_SPEED, EPS);
        // And explicit numeric values to avoid self-referential checks
        assertEquals(11.0, Constants.MIN_BULLET_SPEED, EPS);
        assertEquals(19.7, Constants.MAX_BULLET_SPEED, EPS);

        assertEquals(1, Constants.ACCELERATION);
        assertEquals(-2, Constants.DECELERATION);
    }
}

File: bot-api/wasm/src/wasmJsTest/kotlin/dev/robocode/tankroyale/botapi/ConstantsTest.kt
package dev.robocode.tankroyale.botapi

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ConstantsTest {

    @Test
    fun givenDefinedConstants_whenChecked_thenValuesMatchSpec() {
        assertEquals(18, Constants.BOUNDING_CIRCLE_RADIUS, "BOUNDING_CIRCLE_RADIUS mismatch")
        assertEquals(1200, Constants.SCAN_RADIUS, "SCAN_RADIUS mismatch")

        assertEquals(10, Constants.MAX_TURN_RATE, "MAX_TURN_RATE mismatch")
        assertEquals(20, Constants.MAX_GUN_TURN_RATE, "MAX_GUN_TURN_RATE mismatch")
        assertEquals(45, Constants.MAX_RADAR_TURN_RATE, "MAX_RADAR_TURN_RATE mismatch")

        assertEquals(8, Constants.MAX_SPEED, "MAX_SPEED mismatch")

        assertEquals(0.1, Constants.MIN_FIREPOWER, 1e-12, "MIN_FIREPOWER mismatch")
        assertEquals(3.0, Constants.MAX_FIREPOWER, 1e-12, "MAX_FIREPOWER mismatch")

        assertEquals(1, Constants.ACCELERATION, "ACCELERATION mismatch")
        assertEquals(-2, Constants.DECELERATION, "DECELERATION mismatch")
    }

    @Test
    fun givenComputedConstants_whenEvaluated_thenMatchFormulasAndSanityChecks() {
        val expectedMinBulletSpeed = 20 - 3 * Constants.MAX_FIREPOWER
        val expectedMaxBulletSpeed = 20 - 3 * Constants.MIN_FIREPOWER

        assertEquals(expectedMinBulletSpeed, Constants.MIN_BULLET_SPEED, 1e-12, "MIN_BULLET_SPEED formula mismatch")
        assertEquals(expectedMaxBulletSpeed, Constants.MAX_BULLET_SPEED, 1e-12, "MAX_BULLET_SPEED formula mismatch")

        // Sanity checks
        assertTrue(Constants.MAX_BULLET_SPEED > Constants.MIN_BULLET_SPEED)
        assertEquals(11.0, Constants.MIN_BULLET_SPEED, 1e-12, "Expected MIN_BULLET_SPEED = 11.0")
        assertEquals(19.7, Constants.MAX_BULLET_SPEED, 1e-12, "Expected MAX_BULLET_SPEED = 19.7")
    }
}
